pkg/config/config.go

```diff
@@ -0,0 +1,159 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"knative.dev/pkg/configmap/informer"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/system"
+)
+
+const (
+	paramBatchMaxDuration  = "batchMaxDuration"
+	paramBatchIdleDuration = "batchIdleDuration"
+)
+
+// these values need to be synced with our templates/configmap.yaml
+var defaultConfigMapData = map[string]string{
+	paramBatchMaxDuration:  "10s",
+	paramBatchIdleDuration: "1s",
+}
+
+type ChangeHandler func(c Config)
+
+type Config interface {
+	// OnChange is used to register a handler to be called when the configuration has been changed
+	OnChange(handler ChangeHandler)
+
+	// BatchMaxDuration returns the maximum batch duration
+	BatchMaxDuration() time.Duration
+	// BatchIdleDuration returns the maximum idle period used to extend a batch duration up to BatchMaxDuration
+	BatchIdleDuration() time.Duration
+}
+type config struct {
+	ctx context.Context
+
+	dataMu            sync.RWMutex
+	batchMaxDuration  time.Duration
+	batchIdleDuration time.Duration
+
+	watcherMu sync.Mutex
+	watchers  []ChangeHandler
+}
+
+func (c *config) BatchMaxDuration() time.Duration {
+	c.dataMu.RLock()
+	defer c.dataMu.RUnlock()
+	return c.batchMaxDuration
+}
+
+func (c *config) BatchIdleDuration() time.Duration {
+	c.dataMu.RLock()
+	defer c.dataMu.RUnlock()
+	return c.batchIdleDuration
+}
+
+func New(ctx context.Context, kubeClient *kubernetes.Clientset, iw *informer.InformedWatcher) (Config, error) {
+	cfg := &config{
+		ctx: ctx,
+	}
+	cm, err := kubeClient.CoreV1().ConfigMaps(system.Namespace()).Get(ctx, "config", metav1.GetOptions{})
```

Comment:
This could conflict if a user deploys karpenter in a namespace with other stuff.  For example, kOps deploys karpenter in the kube-system namespace: https://github.com/kubernetes/kops/blob/e59a6003ff204464509410e70e9fe1a9db9d52a8/upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template#L446

Reply:
We need to scope down our RBAC policies, so I think we'd be better off with a consistent name. Prefixing with `karpenter` should be enough to avoid conflicts. e.g. `karpenter-config-logging`, `karpenter-config-global-settings`

Reply:
SGTM